Timeline that displays PDF files

FelipeDefensor opened the pull request.

FelipeDefensor commented:
@azfoo, could you take a look a this? I think you're familiar enough with the codebase that you might be able to spot something I have missed. Don't worry if you can't understand all of it.

azfoo commented:
One potential issue is that it is currently possible to add a page number greater than the number of pages in the pdf document. The viewer defaults to page 1 for any number greater than the last page. Should this be allowed?

The other is that changing the height doesn't do anything for this timeline. The icons get cropped if the timeline is shorter than the default 30, but don't grow if the height exceeds that. I wonder if there is any functionality in changing the height of this timeline at all.

Nice work btw!

azfoo reviewed (changes requested):
Nice work!

Here are a few other things that might need looking into:
1. The timeline continuously increments page numbers from the last highest page number input. This results in the page numbers increasing beyond the last page of the pdf. There isn't any error thrown, but the viewer remains on the first page. Should this be an error instead of having it happen silently?
2. Changing height on this timeline doesn't really do anything. The icons get cropped if the height is changed to less than the default 30 and becomes slightly unclickable. I guess the other question would be if there there is a need for the user to change height on this timeline in the first place .
3. Attempting to add this timeline without any media causes a crash. 
4. On the side facing the user, I think pdf should be either standardised as pdf or PDF, instead of having Pdf thrown in the mix too.

FelipeDefensor commented on tilia/timelines/pdf/components.py:
```diff
@@ -0,0 +1,58 @@
+
+from tilia.timelines.base.component import TimelineComponent
+
+logger = logging.getLogger(__name__)
```
Nope. Will delete that.

azfoo commented on tilia/ui/menus.py:
```diff
@@ -97,6 +97,7 @@ class AddTimelinesMenu(TiliaMenu):
         (MenuItemKind.ACTION, TiliaAction.TIMELINES_ADD_HARMONY_TIMELINE),
         (MenuItemKind.ACTION, TiliaAction.TIMELINES_ADD_HIERARCHY_TIMELINE),
         (MenuItemKind.ACTION, TiliaAction.TIMELINES_ADD_MARKER_TIMELINE),
+        (MenuItemKind.ACTION, TiliaAction.TIMELINES_ADD_PDF_TIMELINE),
         (MenuItemKind.ACTION, TiliaAction.TIMELINES_ADD_AUDIOWAVE_TIMELINE)
```
Would it also be possible to order the list alphabetically while you're at it?

FelipeDefensor replied:
Good call, let's try and keep all multi-line lists sorted alphabetically. I'd rather do that on a separate commit, though, I had problems merging commits that mixed sorting and altering before.

FelipeDefensor commented:
@azfoo fixed what you mentioned, added a bunch of things I forgot and also fixed some bugs that revealed themselves in the process. I will squash everything related to the PDF timeline into a single commit before merging.

azfoo commented:
![image](https://github.com/TimeLineAnnotator/desktop/assets/45888544/44e4da7e-dfee-4dcf-a842-33a4eeac6874)
I don't know if the page number on the UI is supposed to be an index or the actual page number. It shows up as the right number on the inspector, but not the UI. Pressing `enter` to open the inspector on the elements with the wrong number causes the spinbox to update the UI to show the right value. 

I'm also wondering if being able to change the source of the PDF would be a necessary function.

FelipeDefensor commented:
> I don't know if the page number on the UI is supposed to be an index or the actual page number. It shows up as the right number on the inspector, but not the UI. Pressing `enter` to open the inspector on the elements with the wrong number causes the spinbox to update the UI to show the right value.

It's supposed to be the actual page number. I tested it again here, and it seems to work. See this:

https://github.com/TimeLineAnnotator/desktop/assets/87732202/21a7a9b0-0d59-4571-845b-722e0d5911fb

FelipeDefensor commented:
> I'm also wondering if being able to change the source of the PDF would be a necessary function.

I thought about that and coudn't think of any use case that wouldn't be solved just as well by creating a new timeline. Change the source PDF on an existing timeline would only be useful if you wanted to preserve the existing pdf markers. But, if you're changing the file, how could the markers still indicate meaningful timestamps on that new file?

It wouldn't be too much trouble implementing that functionality though, with what we already have. Do you think it's worth it?

azfoo commented:
> > I don't know if the page number on the UI is supposed to be an index or the actual page number. It shows up as the right number on the inspector, but not the UI. Pressing `enter` to open the inspector on the elements with the wrong number causes the spinbox to update the UI to show the right value.
> 
> It's supposed to be the actual page number. I tested it again here, and it seems to work. See this:

If you add a marker, and set it to the last page, then add another marker at a time after that, the UI still shows the new marker to be a higher page than the last possible page.

azfoo commented:
> > I'm also wondering if being able to change the source of the PDF would be a necessary function.
> 
> It wouldn't be too much trouble implementing that functionality though, with what we already have. Do you think it's worth it?

Good point. I guess it isn't necessary then.

azfoo commented:
Two more points:
- I don't know if the delete button is necessary, considering all other components can be ordinarily deleted with delete on the keyboard.
- While it isn't possible to close the viewer from the window itself, closing the viewer from the OS makes it impossible to reopen without restarting the app. Should there be a different way to reopen the window?

FelipeDefensor commented:
> Two more points:
> 
>     * I don't know if the delete button is necessary, considering all other components can be ordinarily deleted with delete on the keyboard.

Good point. We should also remove the analogous button on the marker timeline.
 
>     * While it isn't possible to close the viewer from the window itself, closing the viewer from the OS makes it impossible to reopen without restarting the app. Should there be a different way to reopen the window?

Yes, I think there should. We have an analogous problem with the video window. Maybe it's time to create a "View" menu where the user can toggle the visibility of those windows. I have created a task in the project with High priority for this.

FelipeDefensor commented:
> > > I don't know if the page number on the UI is supposed to be an index or the actual page number. It shows up as the right number on the inspector, but not the UI. Pressing `enter` to open the inspector on the elements with the wrong number causes the spinbox to update the UI to show the right value.
> > 
> > 
> > It's supposed to be the actual page number. I tested it again here, and it seems to work. See this:
> 
> If you add a marker, and set it to the last page, then add another marker at a time after that, the UI still shows the new marker to be a higher page than the last possible page.

That would be the kind of thing that would get me very frustrated once I found it weeks later. Glad I've got another pair of eyes to spot it.

FelipeDefensor commented:
Well, I think that's all of it, for now.

azfoo reviewed (commented):
Could you also remove the deletes from the list in TiliaAction?

And I'm not sure if its a me thing again, but the pdf tests are failing.

FelipeDefensor commented:
> Could you also remove the deletes from the list in TiliaAction?
Actually, no, they are used by the right-click menus.
 
> And I'm not sure if its a me thing again, but the pdf tests are failing.
Not a you thing at all. Fixed them.

azfoo commented:
Yay! I think that's all done then.
